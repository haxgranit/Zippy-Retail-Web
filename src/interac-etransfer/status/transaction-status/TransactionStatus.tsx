import {
  Suspense, useEffect, useState, lazy,
} from 'react';
import {
  Col, Row,
} from 'react-bootstrap';
import { useNavigate, useParams } from 'react-router-dom';
import { useIsAuthenticated, useMsal } from '@azure/msal-react';
import Breadcrumbs, { Crumb } from '../../../common/Breadcrumbs';
import Api, { Transaction } from '../../../api';
import { useAppSelector } from '../../../app/hooks';
import { selectUser, UserState } from '../../../features/user/userSlice';
import QuickLinks from '../components/QuickLinks';
import CommonPageContainer from '../../../common/CommonPageContainer';
import { TransactionTypeEnum } from './TransactionTypeEnum';
import { TransactionStatusEnum } from './TransactionStatusEnum';

export interface TransactionInterface {
  type: TransactionTypeEnum;
  status: TransactionStatusEnum;
  id: string;
}

export interface TransactionProps {
  user: any;
  transaction: Transaction | undefined;
}

export default function TransactionStatus() {
  const isAuthenticated = useIsAuthenticated();
  const navigate = useNavigate();
  const { instance, accounts } = useMsal();
  const { type, status, id } = useParams<Partial<TransactionInterface>>();
  const [transaction, setTransaction] = useState<Transaction | undefined>(undefined);
  const [userState, setUserState] = useState<UserState>({
    firstName: null,
    lastName: null,
    email: null,
  } as UserState | any);
  const [crumbs, setCrumbs] = useState<Array<Crumb>>([
    { label: 'Status', link: '/interac-etransfer/status' },
  ] as Array<Crumb>);
  let user: any;
  if (isAuthenticated) {
    ({ user } = useAppSelector(selectUser));
  } else {
    ({ user } = {
      user: {
        firstName: null,
        lastName: null,
        email: null,
      } as any,
    } as unknown as UserState);
  }

  useEffect(() => {
    new Api(instance, accounts[0])
      .getInteracEtransferTransaction(Number(id))
      .then((data) => setTransaction(data));

    const statusLink = `/interac-etransfer/status/${type}`;
    let typeLabel = '';
    let crumbsTemp: Array<Crumb> = crumbs;
    switch (type) {
      case TransactionTypeEnum.SENT:
        crumbsTemp = [...crumbsTemp, { label: 'Sent', link: statusLink }];
        typeLabel = 'Send Money';
        break;
      case TransactionTypeEnum.REQUESTED:
        crumbsTemp = [...crumbsTemp, { label: 'Requested', link: statusLink }];
        typeLabel = 'Request Money';
        break;
      case TransactionTypeEnum.RECEIVED:
        crumbsTemp = [...crumbsTemp, { label: 'Received', link: statusLink }];
        typeLabel = 'Receive Money';
        break;
      default:
        navigate(`/interac-etransfer/status/${type}`);
    }

    switch (status) {
      case TransactionStatusEnum.PENDING:
        crumbsTemp = [...crumbsTemp, { label: `${typeLabel} Pending` }];
        break;
      case TransactionStatusEnum.CANCELED:
        crumbsTemp = [...crumbsTemp, { label: `${typeLabel} Canceled` }];
        break;
      case TransactionStatusEnum.COMPLETED:
        crumbsTemp = [...crumbsTemp, { label: `${typeLabel} Completed` }];
        break;
      case TransactionStatusEnum.REMINDER:
        crumbsTemp = [...crumbsTemp, { label: `${typeLabel} Reminder` }];
        break;
      default:
        navigate(`/interac-etransfer/status/${type}`);
    }

    setCrumbs(crumbsTemp);
    setUserState(user);
  }, []);

  const TransactionComponent = lazy(() => import(`./${type}/${status}/${type?.replace(/^./, (str) => str.toUpperCase())}${status?.replace(/^./, (str) => str.toUpperCase())}`));

  return (
    <>
      <CommonPageContainer title="Status">
        <Row>
          <Col xs={9} className="transaction-details">
            <Breadcrumbs
              crumbs={crumbs}
            />
            <h2>{crumbs[2]?.label}</h2>
            <div className="details">
              <Suspense fallback={<div>Loading...</div>}>
                <TransactionComponent
                  transaction={transaction}
                  user={userState}
                />
              </Suspense>
            </div>
          </Col>
          <Col xs={3}>
            <QuickLinks />
            <div className="quick-tips">
              <h4>Your Interac e-Transfer Details</h4>
              <ul>
                <li>
                  You can also view information about your InteracÂ® e-Transfer on the
                  &quot;Account Details&quot; page. Select the account from which you made the
                  transfer to view the details of your transfer in the list of
                  transactions
                </li>
              </ul>
            </div>
          </Col>
        </Row>
      </CommonPageContainer>
    </>
  );
}
