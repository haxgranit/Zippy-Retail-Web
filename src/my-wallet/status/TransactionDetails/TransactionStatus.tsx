import {
  Suspense, useEffect, useState, lazy,
} from 'react';
import { useLocation, useNavigate, useParams } from 'react-router-dom';
import { useIsAuthenticated, useMsal } from '@azure/msal-react';
import Api from '../../../api';
import { useAppSelector } from '../../../app/hooks';
import { selectUser, UserState } from '../../../features/user/userSlice';
import { TransactionTypeEnum } from './TransactionTypeEnum';
import PageContainer from '../../../common/PageContainer';
import { Transaction } from '../../../constants/type/Transaction';
import { TransactionStatusEnum } from '../../../constants/enum/TransactionStatusEnum';
import { TransactionTypePastTenseEnum } from '../../../constants/enum/TransactionTypePastTenseEnum';

export interface TransactionInterface {
  type: TransactionTypeEnum;
  status: TransactionStatusEnum;
  id: string;
}

export interface TransactionProps {
  type: TransactionTypePastTenseEnum;
  user: any;
  transaction: Transaction | undefined;
  setCurrentStatus: any;
}

export default function TransactionStatus() {
  const navigate = useNavigate();
  const isAuthenticated = useIsAuthenticated();
  const { instance, accounts } = useMsal();
  const { type, status, id } = useParams<Partial<TransactionInterface>>();
  const { state = { type } as any } = useLocation();
  const [currentType] = useState<TransactionTypeEnum>((state?.type || type) as TransactionTypeEnum);
  const [
    currentStatus,
  ] = useState<TransactionStatusEnum>(status as TransactionStatusEnum);
  const [transaction, setTransaction] = useState<Transaction | undefined>(undefined);
  const [userState, setUserState] = useState<UserState>({
    firstName: null,
    lastName: null,
    email: null,
  } as UserState | any);
  let user: any;
  if (isAuthenticated) {
    ({ user } = useAppSelector(selectUser));
  } else {
    ({ user } = {
      user: {
        firstName: null,
        lastName: null,
        email: null,
      } as any,
    } as unknown as UserState);
  }

  useEffect(() => {
    new Api(instance, accounts[0])
      .getTransfer(Number(id))
      .then((data) => setTransaction(data));
    setUserState(user);
  }, [instance, accounts, user]);

  const TransactionComponent = lazy(() => import(`./${type}/${status}/${type?.replace(/^./, (str) => str.toUpperCase())}${status?.replace(/^./, (str) => str.toUpperCase())}`));

  return (
    <>
      <PageContainer
        title="Transactions History"
        subTitle="Made Fun With Zippy!"
        className="status-list"
        backdropImage="backdrop-image-3"
        showClose
        closeHandler={() => navigate(`/my-wallet/transaction-history/${currentType}`)}
      >
        <div className="details">
          <Suspense fallback={<div>Loading...</div>}>
            { currentType && currentStatus
            && (
              <TransactionComponent
                transaction={transaction}
                user={userState}
                type={type}
              />
            )}
          </Suspense>
        </div>
      </PageContainer>
    </>
  );
}
