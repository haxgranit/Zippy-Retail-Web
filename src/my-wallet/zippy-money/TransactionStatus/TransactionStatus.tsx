import {
  Button,
} from 'react-bootstrap';
import { useNavigate } from 'react-router-dom';
import PageContainer from '../../../common/PageContainer';
import { TransactionInterface } from '../../../constants/interface/TransactionInterface';

export default function TransactionStatus({
  transaction,
  transactionType,
  resetMainInfo,
}: Pick<TransactionInterface, 'transaction' | 'transactionType' | 'resetMainInfo'>) {
  const navigate = useNavigate();

  return (
    <>
      <PageContainer title="Personal Account" subTitle="Made Fun With Zippy!" className="transaction-status">
        <div className="title">
          Transaction Status
        </div>
        <div className="status">
          <div><i className="zippy-cash-icon zc-verified" /></div>
          <div>
            Your $100 transfer has been sent to Duane Tough.
          </div>
        </div>
        <div className="action">
          <Button
            className="zippy-btn"
            onClick={resetMainInfo}
          >
            Send Another Transfer
          </Button>
          <Button
            className="zippy-btn btn-neutral"
            onClick={() => navigate(`/my-wallet/zippy-money/${transactionType}/transaction-details/${transaction?.id}`)}
          >
            View Transfer Details
          </Button>
        </div>
      </PageContainer>
    </>
  );
}
