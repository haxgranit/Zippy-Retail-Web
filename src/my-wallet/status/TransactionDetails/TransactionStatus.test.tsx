import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import TransactionStatus from './TransactionStatus';

const ReactTestRenderer = require('react-test-renderer');

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}));

describe('TransactionStatus Component', () => {
  it('matches the snapshot', () => {
    const tree = ReactTestRenderer.create(
      <BrowserRouter>
        <TransactionStatus />
      </BrowserRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
