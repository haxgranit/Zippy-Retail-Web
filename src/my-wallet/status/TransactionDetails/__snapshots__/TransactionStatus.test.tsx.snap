// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionStatus Component matches the snapshot 1`] = `
<div
  className="status-list"
>
  <div
    className="zippy-cash-container-wrap"
  >
    <div
      className="backdrop backdrop-image-3"
    />
    <div
      className="zippy-cash-container"
    >
      <div
        className="zippy-cash-container-header"
      >
        <i
          aria-label="close"
          className="zippy-cash-icon zc-cross"
          onClick={[Function]}
          onKeyUp={[Function]}
          role="button"
          tabIndex={0}
        />
        <h1
          className="title"
        >
          Transactions History
        </h1>
        <h2
          className="catch"
        >
          Made Fun With Zippy!
        </h2>
      </div>
      <div
        className="zippy-cash-container-body"
      >
        <div
          className="details"
        />
      </div>
    </div>
  </div>
</div>
`;
